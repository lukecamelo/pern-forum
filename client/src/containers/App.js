import React, { Component } from 'react'
import styled from 'styled-components'
import { Provider } from 'react-redux'
import store from '../store'

import Login from '../components/Login'
import Signup from '../components/Signup'

const Wrapper = styled.div`
  text-align: center;
`
const Header = styled.div`
  height: 150px;
  padding: 20px;
  color: #564154;
`

class App extends Component {
  state = {
    user: null,
    isLoggedIn: false
  }

  // componentDidMount = () => {
  //   if (localStorage.Authorization) {
  //     this.getUser()
  //   }
  // }

  mockLogin = () => {
    fetch('/auth/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ username: 'rediscover', password: 'testpassword' })
    })
      .then(res => res.json())
      .then(res => {
        localStorage.Authorization = res.token
        this.setState({
          isLoggedIn: true
        })
      })
  }

  getUser = () => {
    fetch('/user/profile', {
      method: 'GET',
      headers: {
        Authorization: 'Bearer ' + localStorage.Authorization
      }
    })
      .then(res => res.json())
      .then(res => {
        this.setState({
          user: res.username
        })
      })
  }

  render() {
    return (
      <Provider store={store}>
        <Wrapper>
          <Header>
            <Login />
            <Signup/>
          </Header>
        </Wrapper>
      </Provider>
    )
  }
}

export default App
