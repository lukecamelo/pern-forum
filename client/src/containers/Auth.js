import React from 'react'
import Login from '../components/Login'
import Signup from '../components/Signup'

import { connect } from 'react-redux'

const Auth = ({ isLoggedIn }) => {
  return <div>{isLoggedIn ? <Login /> : <Signup />}</div>
}

const mapStateToProps = state => ({
  isLoggedIn: state.auth.isLoggedIn
})

export default connect(mapStateToProps)(Auth)
