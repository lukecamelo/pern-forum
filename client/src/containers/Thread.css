blockquote {
  margin-left: 0;
  padding: 0 1em;
  color: #3f83c7;
  border-left: 0.25em solid #0266c8;
}
