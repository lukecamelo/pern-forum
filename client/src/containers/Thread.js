import React, { Component } from 'react'
import styled from 'styled-components'
import axios from 'axios'
import { Container } from '../components/Login'
import NavBar from '../components/NavBar'
import PostForm from '../components/PostForm'
import Pagination from '../components/Pagination'
import PostList from '../components/PostList'
import Avatar from 'react-avatar'
// import { Author, PostWrapper, PostContent } from '../components/PostList'

const ThreadWrapper = styled.section`
  background-color: #dff4ff;
  color: #564154;
  padding: 2em;
`
const StyledThread = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  color: #32343b;
  background-color: white;
  margin-bottom: 1em;
  border: 2px solid ${props => props.theme.primary};
`
export const ThreadHeader = styled.h2`
  margin-left: 1em;
  border-bottom: 2px solid ${props => props.theme.secondary}
  text-align: left;
`
const ThreadContent = styled.p`
  padding-left: 10px;
  padding-right: 10px;
`
const PostWrapper = styled.div`
  // border: 1px solid ${props => props.theme.secondary};
  display: flex;
  background-color: white;
  text-align: left;
  padding: 0 1em 1em 1em;
`
const User = styled.div``
const Author = styled.h2`
  color: ${props => props.theme.primary}
  font-size: 18px;
`
const PostContent = styled.p`
  color: black;
  font-size: 16px;
  padding: 0 15px 15px 15px;
`

export class Thread extends Component {
  state = {
    title: '',
    content: '',
    author: '',
    threadPosts: [],
    userId: null,
    threadHasLoaded: false
  }

  componentDidMount = async () => {
    const {
      match: { params }
    } = this.props

    this.fetchThreadAndAuthor()
  }

  fetchSingleThread = async threadId => {
    const thread = await axios.get(`/api/threads/${threadId}`)
    console.log(thread.data)
    return thread.data
  }

  fetchThreadAuthor = async userId => {
    const author = await axios.get(`/api/users/${userId}`)
    return author.data
  }

  fetchThreadAndAuthor = async () => {
    const thread = await this.fetchSingleThread(this.props.match.params.id)
    console.log(thread)
    const author = await this.fetchThreadAuthor(thread.userId)
    this.setState({
      title: thread.title,
      content: thread.content,
      threadPosts: thread.Post,
      userId: thread.userId,
      author: author.username,
      threadHasLoaded: true
    })
  }
  render() {
    const { title, content, author, threadHasLoaded, threadPosts } = this.state
    if (threadHasLoaded) {
      const posts = threadPosts.map(post => (
        <PostWrapper key={post.id}>
          <User>
            <Author>{post.author}</Author>
            <Avatar size="150" src={post.user.avatarUrl} />
          </User>
          <PostContent>{post.content}</PostContent>
        </PostWrapper>
      ))
      const op = posts.shift()

      return (
        <Container>
          <NavBar />
          <ThreadWrapper>
            <ThreadHeader>
              {title}, posted by {author}
            </ThreadHeader>
            <StyledThread>{op}</StyledThread>

            {posts.length ? (
              <Pagination data={posts}>
                <PostList />
              </Pagination>
            ) : (
              <h1>make the first post!</h1>
            )}
          </ThreadWrapper>
          <PostForm threadId={this.props.match.params.id} />
        </Container>
      )
    } else {
      return (
        <Container>
          <NavBar />
          <h1>Loading thread...</h1>
        </Container>
      )
    }
  }
}

export default Thread
