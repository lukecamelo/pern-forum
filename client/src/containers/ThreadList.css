.link {
  display: flex;
  box-sizing: border-box;
  color: #2e262e;
  text-decoration: none;
  padding: 1em 2em;
  font-size: 2em;
  border-bottom: 2px solid #0266c8;
  width: 100%;
  text-align: left;
}

.link:hover {
  color: white;
  background-color: #b28bc0;
}

.title {
  display: flex;
  text-decoration: none;
  color: #3d4852; 
  padding: .5em 1em 0 1em;
  width: 100%;
  font-size: 1.5em;
  justify-self: flex-start;
  align-self: center;
  font-family: 'Roboto', sans-serif;
  /* font-weight: 300; */
}

.title-pages {
  display: flex;
  flex-direction: column;
}

.page-controls {
  padding: 0 0 0 1.3em;
}

.page-button {
  display: inline-block;
  text-decoration: none;
  color: #0266c8;
  padding: 0.5em 0.75em;
  transition: .2s;
}

.page-button:hover {
  color: #fff;
  background-color: #0266c8;
}

.title:hover {
  color: #00a8ff;
}

.author {
  display: flex;
  flex-direction: row;
  padding: .5em 1em;
}

.item {
  display: flex;
  text-align: center;
  flex-direction: column;
  padding: 0 .5em 0 .5em;
  color: #00a8ff;
  font-family: 'Lato', sans-serif;
  font-weight: 300;
  font-style: italic;
}

.item p.username {
  margin: 0;
  width: 100px;
  text-overflow: ellipsis;
  overflow: hidden;
}

.item p.tag {
  text-transform: uppercase;
  letter-spacing: 2px;
  margin: 0;
  color: #2e262e;
}

.markdown-shiz {
  padding-top: 2em;
}

.markdown-shiz p {
  margin: 0;
}

@media screen and (max-width: 968px) {
  .link {
    font-size: 1em;
  }
  .title {
    font-size: 1em;
  }
  .page-controls {
    padding-left: .5em;
  }
  .page-button {
    font-size: 1em;
  }
  .item {
    display: none;
  }
  .thread-author {
    display: flex;
  }
}

@media screen and (max-width: 532px) {
  .markdown-shiz {
    padding-top: 0em;
  }
}