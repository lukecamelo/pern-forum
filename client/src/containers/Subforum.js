import React, { Component } from 'react'
import { connect } from 'react-redux'
import { fetchSubforumThreads, fetchData } from '../actions/threadActions'
import ThreadList from './ThreadList'
import Pagination from '../components/Pagination'
import { Banner } from './App'
import { H1, SubforumLink } from '../styled'
import NavBar from '../components/NavBar'
import Loader from '../components/Loader'

class Subforum extends Component {
  state = {
    hasLoaded: false,
    subforumName: ''
  }

  async componentDidMount() {
    await this.props.fetchData()
    switch (this.props.match.params.id) {
      case '1':
        this.setState({ subforumName: 'General Discussion' })
        break
      case '2':
        this.setState({ subforumName: 'Video Games' })
        break
      default:
        break
    }
    this.setState({
      hasLoaded: true
    })
  }

  render() {
    if (this.state.hasLoaded) {
      return (
        <React.Fragment>
          <NavBar />
          <Banner>
            <div>
              <SubforumLink
                to="/subforums"
                style={{ margin: '0 auto', color: 'white' }}
              >
                Forums
              </SubforumLink>{' '}
              <H1 style={{ margin: '0 auto', color: 'white', display: 'inline' }}> -> </H1>{' '}
              <H1 style={{ margin: '0 auto', color: 'white', display: 'inline' }}>
                {this.state.subforumName}
              </H1>
            </div>
          </Banner>
          <Pagination
            data={this.props.threads}
            currentPage={this.props.match.params.page}
            context="threads"
            pageSize={15}
            isLoggedIn={this.props.auth.isLoggedIn}
            subforumId={this.props.match.params.id}
          >
            {data => (
              <ThreadList data={data} subforumId={this.props.match.params.id} />
            )}
          </Pagination>
        </React.Fragment>
      )
    } else {
      return <Loader />
    }
  }
}

const mapStateToProps = state => ({
  threads: state.threadData.threads,
  auth: state.auth
})

export default connect(
  mapStateToProps,
  { fetchSubforumThreads, fetchData }
)(Subforum)
