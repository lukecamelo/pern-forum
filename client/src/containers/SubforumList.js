import React, { Component } from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'

import api from '../services/api'
import moment from 'moment'

import { Banner } from './App'
import { H1 } from '../styled'
import '../css/SubforumList.css'
import { ThreadLink } from './ThreadList'
import Loader from '../components/Loader'

const SubforumContainer = styled.div`
  width: 75vw;
  margin: 0 auto 2em auto;
  background-color: white;
  box-shadow: ${({ theme }) => theme.mediumShadow};
  @media screen and (max-width: 700px) {
    width: 90vw;
  }
`
const LatestThread = styled(Link)`
  text-decoration: none;
  color: ${({ theme }) => theme.primary};
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0;
  white-space: nowrap;
  display: block;
  width: 250px;
  transition: 0.2s;
  &:hover {
    color: ${({ theme }) => theme.secondary};
    text-decoration: underline;
  }
`
const InfoText = styled.p`
  margin: 0;
  color: rgb(61, 72, 82);
  font-size: 18px;
`
/*
 ************* 
 This whole component is a big mess
 TODO: fix the mess
 ************* 
*/

class SubforumList extends Component {
  state = {
    latestGeneral: '...',
    latestGames: '...',
    hasLoaded: false
  }

  componentDidMount = async () => {
    let latestGeneral = await api.threads.getLatestSubforumThread(1)
    let latestGames = await api.threads.getLatestSubforumThread(2)
    this.setState({
      latestGeneral,
      latestGames,
      hasLoaded: true
    })
  }

  render() {
    const { hasLoaded, latestGeneral, latestGames } = this.state
    if (hasLoaded) {
      return (
        <React.Fragment>
          <Banner>
            <H1 style={{ margin: '0 auto', color: 'white' }}>Forums</H1>
          </Banner>
          <SubforumContainer
            className="subforums"
            style={{ overflow: 'hidden' }}
          >
            <ThreadLink style={{ fontSize: '1.5em' }}>
              <div className="title-pages">
                <Link className="title" to="/subforum/1/page/1">
                  General Discussion
                </Link>
              </div>

              <div
                className="latest-thread"
                style={{
                  justifyContent: 'center',
                  padding: '0.5em 1em',
                  width: '200px',
                  textOverflow: 'ellipsis'
                }}
              >
                <div
                  className="item thread-author"
                  style={{ textAlign: 'left' }}
                >
                  <LatestThread
                    to={`/thread/${this.state.latestGeneral.id}/page/1`}
                  >
                    {latestGeneral.title}
                  </LatestThread>
                  <InfoText>By {latestGeneral.Post[0].author}</InfoText>
                  <InfoText>
                    {moment(latestGeneral.Post[0].createdAt).fromNow()}
                  </InfoText>
                </div>
              </div>
            </ThreadLink>

            <ThreadLink style={{ fontSize: '1.5em' }}>
              <div className="title-pages">
                <Link className="title" to="/subforum/2/page/1">
                  Video Games
                </Link>
              </div>
              <div
                className="latest-thread"
                style={{
                  justifyContent: 'center',
                  padding: '0.5em 1em',
                  width: '200px',
                  textOverflow: 'ellipsis'
                }}
              >
                <div
                  className="item thread-author"
                  style={{ textAlign: 'left' }}
                >
                  <LatestThread
                    to={`/thread/${this.state.latestGames.id}/page/1`}
                  >
                    {latestGames.title}
                  </LatestThread>
                  <InfoText>By {latestGames.Post[0].author}</InfoText>
                  <InfoText>
                    {moment(latestGames.Post[0].createdAt).fromNow()}
                  </InfoText>
                </div>
              </div>
            </ThreadLink>
          </SubforumContainer>
        </React.Fragment>
      )
    } else {
      return (
        <React.Fragment>
          <Loader />
        </React.Fragment>
      )
    }
  }
}

export default SubforumList
