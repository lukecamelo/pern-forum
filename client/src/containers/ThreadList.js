import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { fetchData, fetchThreads } from '../actions/threadActions'
import { filterAuthor, checkForPosts } from '../utils/threadHelpers'
import Loader from '../components/Loader'
import styled from 'styled-components'
import '../css/ThreadList.css'

export const ListWrapper = styled.main`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 75%;
  margin: 0 auto;
  box-shadow: ${props => props.theme.largeShadow};
  @media screen and (max-width: 700px) {
    width: 90%;
  }
`
const ThreadLink = styled.div`
  background-color: white;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  flex-grow: 1;
  width: 100%;
  border-bottom: solid 1px hsla(270, 7%, 92%, 1);
  transition: 0.2s;
`

export class ThreadList extends Component {
  state = {
    hasLoaded: false
  }

  async componentDidMount() {
    await this.props.fetchData()
    await this.props.fetchThreads()
    this.setState({
      hasLoaded: true
    })
  }

  showPageNumbers = thread => {
    let pageCount = parseInt(thread.Post.length / 10, 10)
    if (thread.Post.length % 10 > 0) {
      pageCount++
    }
    let controls = []
    for (let i = 1; i <= pageCount; i++) {
      controls.push(
        <Link
          to={`/thread/${thread.id}/page/${i}`}
          key={i}
          className="page-button"
        >
          {i}
        </Link>
      )
    }
    return controls
  }

  render() {
    const { data: threads, users } = this.props
    if (this.state.hasLoaded && checkForPosts(threads)) {
      const threadLinks = threads.map(thread => {
        return (
          <ThreadLink key={thread.id}>
            <div className="title-pages">
              <Link className="title" to={`/thread/${thread.id}/page/1`}>
                {thread.title}
              </Link>
              <div className="page-controls">
                {thread.Post.length > 10 ? this.showPageNumbers(thread) : null}
              </div>
            </div>
            <div className="author">
              <div className="item thread-author">
                <p className='tag'>Author</p>
                <p className="username">
                  {filterAuthor(users, thread.userId)}
                </p>
              </div>

              <div className="item">
                <p className='tag'>Posts</p>
                {thread.Post.length}
              </div>

              <div className="item">
                <p className='tag'>Latest</p>
                <p className="username">{thread.Post[0].author}</p>
              </div>
            </div>
          </ThreadLink>
        )
      })
      return (
        <React.Fragment>
          <ListWrapper>{threadLinks}</ListWrapper>
        </React.Fragment>
      )
    } else {
      return <Loader />
    }
  }
}

const mapStateToProps = state => ({
  users: state.threadData.users
})

export default connect(
  mapStateToProps,
  { fetchData, fetchThreads }
)(ThreadList)
