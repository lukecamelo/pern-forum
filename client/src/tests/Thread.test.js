import React from 'react'
import { Thread, ThreadHeader } from '../containers/Thread'
import { shallow } from 'enzyme'

describe('<Thread />', () => {
  it('fetches the thread', async () => {
    const wrapper = shallow(<Thread match={{ params: { id: 1 } }} />)
    expect(wrapper.find(ThreadHeader).length).toEqual(1)
  })
})
