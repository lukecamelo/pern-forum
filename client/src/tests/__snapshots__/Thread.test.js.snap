// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Thread /> renders thread content 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Thread
    match={
        Object {
            "params": Object {
              "id": 1,
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Connect(NavBar) />,
        <h1>
          Loading thread...
</h1>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Loading thread...",
        },
        "ref": null,
        "rendered": "Loading thread...",
        "type": "h1",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Connect(NavBar) />,
          <h1>
            Loading thread...
</h1>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Loading thread...",
          },
          "ref": null,
          "rendered": "Loading thread...",
          "type": "h1",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
